### DISCUSSION TITLE
What are the rules for user defined types for providing __sig in js library

### DISCUSSION THREAD

**Discussion by k1nder10** [emscripten-core/emscripten#17146] [Q&A]
I have a js library. I use mergeInto() in it. I have to describe function signature for emscripten compiler but I don't know the rules for user-defined types. 

void foo(myClass1 m1, myClass2 m2, emscripten::EM_VAL); 
foo__sig: 'v???'

**Comment by kripken** [comment]
Passing classes by value is tricky. The JS library API is very low-level, so you'd need to find out how the C++ ABI for such classes ends up implemented in wasm imports. That might not be easy and is not recommended. Instead, using embind at a high level is probably best.

JS library methods are good when your types are simple:
* `v` - void (no returned value)
* `i` - i32
* `j` - i64
* `f` - f32
* `d` - f64
* `p` - pointer (usually i32, but i64 in wasm64)

**Comment by k1nder10** [comment]
What about std::string? Can I use 'p' sign for it?

**Reply by k1nder10** [reply]
I mean is there an exception for std::string? I guess it would be pretty useful to have one.

**Reply by kripken** [reply]
For "higher-level" types like `std::string` it is best to use [embind](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html).

**Reply by k1nder10** [reply]
Can you have a look at the convertJsFunctionToWasm() in the out .js file produced by emcc? There's no check for 'p' at all. So, if one uses it, the error occurs.

The code from that function:
 if (typeof WebAssembly.Function == "function") {
    var typeNames = {
      'i': 'i32',
      'j': 'i64',
      'f': 'f32',
      'd': 'f64'
    };
    var type = {
      parameters: [],
      results: sig[0] == 'v' ? [] : [typeNames[sig[0]]]
    };

**Reply by k1nder10** [reply]
To enable using of 'p' I added it to typeCodes&typeNames objects in the out .js file though it has no difference between 'i'. It seems to be working. Can it break sth in the future? Are there any pitfalls? 
  
  var typeCodes = {
    'p': 0x7f, // i32
    'i': 0x7f, // i32
    'j': 0x7e, // i64
    'f': 0x7d, // f32
    'd': 0x7c, // f64
  };

**Reply by kripken** [reply]
Atm we use `p` internally in JS library function signatures, but not in runtime stuff like `addFunction`, as you ran into. I don't think we have plans to change that. But you are modifying internals in ways that might change. It would be safer to use embind.

**Reply by sbc100** [reply]
convertJsFunctionToWasm() should always contain a `p` handler.  If yours doesn't maybe you are using an older version of emscripten?:

https://github.com/emscripten-core/emscripten/blob/8818da2a1f7b70c32698f7f40ebe72d4c4dc86b3/src/runtime_functions.js#L68-L78

https://github.com/emscripten-core/emscripten/blob/8818da2a1f7b70c32698f7f40ebe72d4c4dc86b3/src/runtime_functions.js#L22-L32

**Reply by k1nder10** [reply]
Yes, I used 3.1.8. Today I updated emsdk to 3.1.13 which supports 'p' for pointers. Thanks!
